File: lib/domain/usecases/repository/create_repository_usecase.dart
import '../../entities/repository.dart';
import '../../repositories/i_repository_repository.dart';

/// 创建仓库用例
class CreateRepositoryUseCase {
  CreateRepositoryUseCase(this._repository);

  final IRepositoryRepository _repository;

  Future<Repository> execute({
    required String name,
    String icon = 'folder',
    String color = '#3B82F6',
  }) async {
    if (name.trim().isEmpty) {
      throw ArgumentError('仓库名称不能为空');
    }

    final repository = Repository.create(
      name: name.trim(),
      icon: icon,
      color: color,
    );

    return await _repository.create(repository);
  }
}

File: web/src/application/usecases/repository/create-repository-usecase.ts
import { Branch } from '../../../domain/entities/branch';
import { Repository } from '../../../domain/entities/repository';
import { IBranchRepository } from '../../../domain/repositories/i-branch-repository';
import { IRepositoryRepository } from '../../../domain/repositories/i-repository-repository';
import { validateNotEmpty } from '../../../core/utils/validators';

export interface CreateRepositoryInput {
  name: string;
  description?: string | null;
  icon?: string;
  color?: string;
}

export class CreateRepositoryUseCase {
  constructor(
    private repositoryRepo: IRepositoryRepository,
    private branchRepo: IBranchRepository
  ) {}

  async execute(input: CreateRepositoryInput): Promise<Repository> {
    validateNotEmpty(input.name, '仓库名称');
    const repo = Repository.create(
      input.name,
      input.icon ?? 'repository',
      input.color ?? '#3B82F6',
      input.description ?? null
    );
    const createdRepo = await this.repositoryRepo.create(repo);
    const mainBranch = Branch.main(createdRepo.id, createdRepo.color);
    await this.branchRepo.create(mainBranch);
    const repoWithDefaultBranch = createdRepo.copyWith({
      defaultBranchId: mainBranch.id,
    });
    return this.repositoryRepo.update(repoWithDefaultBranch);
  }
}
